Kernel/Arch/i686/Bootloader/stivale2.c

```diff
@@ -0,0 +1,314 @@
+    pagesDirectoryEntry->readWrite = 1;
+    pagesDirectoryEntry->tableAddr = (uint32_t)&kernelPageTable[1] >> PAGE_TABLE_ENTRY_SHIFT;
+
+    for (uintptr_t addr = KERNEL_START; addr < KERNEL_END; addr += ARCH_PAGE_SIZE) {
+        size_t kernelMemoryIdx = addr >> PAGE_TABLE_ENTRY_SHIFT;
+        // OR the mask with 0x400 so that pagesDirectoryEntry is used after kernelDirectoryEntry.
+        // This only works because these two pages tables are next to each other in memory.
+        struct TableEntry* kernelMemoryTableEntry = &kernelPageTable[0].pages[kernelMemoryIdx & (0x400 | PAGE_TABLE_ENTRY_MASK)];
+        kernelMemoryTableEntry->present = 1;
+        kernelMemoryTableEntry->readWrite = 1;
+        kernelMemoryTableEntry->frame = kernelMemoryIdx - (KERNEL_BASE >> PAGE_TABLE_ENTRY_SHIFT);
+    }
+}
+
+/**
+ * @brief identity map from 0x00000000 -> LOWMEM_END
+ *
+ */
+__attribute__((section(".early_text")))
+static void stage1MapLowMemory(void)
+{
+    // No need to get an index here since we're starting at 0x00000000,
+    // which corresponds to the first entry in the page directory.
+    // WARNING: code assumes that the kernel won't be greater than 3MB
+    struct DirectoryEntry* lowMem = &pageDirectory.entries[0];
+    lowMem->present = 1;
+    lowMem->readWrite = 1;
+    lowMem->tableAddr = (uint32_t)&lowMemoryPageTable >> PAGE_TABLE_ENTRY_SHIFT;
+
+    // Map in the entirety of low-memory and stage1
+    for (uintptr_t addr = ARCH_PAGE_SIZE; addr < EARLY_KERNEL_END; addr += ARCH_PAGE_SIZE) {
+        size_t pageIdx = addr >> PAGE_TABLE_ENTRY_SHIFT;
+        struct TableEntry* lowMemoryTableEntry = &lowMemoryPageTable.pages[pageIdx & PAGE_TABLE_ENTRY_MASK];
+        lowMemoryTableEntry->present = 1;
+        lowMemoryTableEntry->readWrite = 1;
+        lowMemoryTableEntry->frame = pageIdx;
+    }
+}
+
+/**
+ * @brief Stivale2 protocol kernel stage 1 entry. Stage 1 is responsible
+ * for providing an entry point for the bootloader in C, performing any
+ * necessary bootstrappign and then calling into the C++ stage 2.
+ *
+ */
+__attribute__((section(".early_text")))
+static void stage1Entry(struct stivale2_struct *info)
+{
+    // By this point the stivale2 bootloader should have provided us a healthy
+    // stack to use (at the location we specified in stivale2_header_i686 struct)
+    if (!info) {
+        // xyris requires information (such as memory maps, framebuffer location, etc)
+        // from the bootloader, so if that's missing we're already in trouble.
+        EarlyPanic("Error: Missing bootloader information!");
+    }
+
+    // preserve bootloader information pointer so that when we move the
+    // stack pointer we don't lose the one thing we need to pass into
+    // stage2 / the higher half stack
+    stivale2Info = info;
+
+    stage1MapLowMemory();
+    stage1MapHighMemory();
+    stage1MapBootloader();
+    stage1EnablePaging();
+    stage1JumpToStage2();
+    EarlyPanic("Error: Execution returned to stage1!");
```

Comment:
I'm curious as to why the compiler doesn't complain that this is unreachable code. Either way it is impossible for execution to ever get here.